server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java

```diff
@@ -264,4 +264,12 @@ protected Set<Long> getSnapshotIdsOnlyInPrimaryStorage(long volumeId) {
 
         return snapshotIdsOnlyInPrimaryStorage;
     }
+
+    public SnapshotInfo convertSnapshotIfNeeded(SnapshotInfo snapshotInfo) {
+        if (snapshotInfo.getParent() == null || !HypervisorType.KVM.equals(snapshotInfo.getHypervisorType())) {
+            return snapshotInfo;
+        }
+
+        return snapshotService.convertSnapshot(snapshotInfo);
```

Comment:
@JoaoJandre, the last time I didn't see this but the converted snapshot is left on the secondary storage and fills the space.

```
[root@lab-cs-test-mgmt export]# ls -latr secondary-a/snapshots/2/1078/
total 10542184
-rw-r--r--. 1 root root 2188509184 28 авг 13,14 c453378b-ac77-4423-9667-5f76273ccd11
-rw-r--r--. 1 root root 2188509184 28 авг 13,17 c25da6f2-636c-4d21-a55d-e2174d0e24c5
-rw-r--r--. 1 root root 2112880640 28 авг 13,18 c25da6f2-636c-4d21-a55d-e2174d0e24c5_temp
drwxr-xr-x. 4 root root         29 28 авг 13,49 ..
-rw-r--r--. 1 root root 2188509184 28 авг 13,54 8db99bcd-3969-4314-8037-ddd2fd9a797e
drwxr-xr-x. 2 root root        236 28 авг 13,54 .
-rw-r--r--. 1 root root 2112880640 28 авг 13,55 8db99bcd-3969-4314-8037-ddd2fd9a797e_temp
```

Reply:
What test did you make to get this result? I'll take a look at it

Reply:
I've just tested this, and could not reproduce. Did you recreate the SSVMs after updating? You need to do so.

Reply:
still the same with recreated SSVMs. I will build packages again and will try again

Reply:
@slavkap In the latest commit I changed the copied snapshots's name to match the original, but that should not matter for this issue, in any case, I have done the following test, in an env with two zones:
1. Create a VM in zone 'A';
2. Create two snapshots, marking them to be copied zone 'B' as well;
Got the following results on storage:
```
root@management:/mnt/sec-nfs# ls -lah snapshots/2/85/
total 3.2G
drwxr-xr-x 2 nobody nogroup 4.0K Aug 28 16:34 .
drwxr-xr-x 6 nobody nogroup 4.0K Aug 28 16:30 ..
-rw-r--r-- 1 nobody nogroup 3.2G Aug 28 16:30 25ced820-28c1-4bee-b41a-b03ebdbd1f76
-rw-r--r-- 1 nobody nogroup 3.0M Aug 28 16:31 57199f60-afd0-424b-9313-9bda9afe34ca
root@management:/mnt/sec-nfs# ls -lah sec-2/snapshots/2/85/
total 6.4G
drwxrwxrwx 2 nobody nogroup 4.0K Aug 28 16:32 .
drwxrwxrwx 4 nobody nogroup 4.0K Aug 28 16:30 ..
-rw-rw-rw- 1 nobody nogroup 3.2G Aug 28 16:31 25ced820-28c1-4bee-b41a-b03ebdbd1f76
-rw-rw-rw- 1 nobody nogroup 3.2G Aug 28 16:32 57199f60-afd0-424b-9313-9bda9afe34ca
```
3. Then I created a template from the second snapshot, I saw that a temporary file was created, but after the process finished the file as deleted, and the storage returned to the above scenario.

Reply:
@JoaoJandre, all good here! Sorry for the trouble!